Enable perf_events to resolve JIT compiled code addresses

brianrob opened the pull request:
Adds a new COMPLUS variable that enables CoreCLR to emit a perfmap file (/tmp/perf-$pid.map).  This file is consumed by perf_events when collecting a trace and is used to resolve code addresses in JIT'd code.

To enable this feature: export COMPlus_PerfMapEnabled=1

stephentoub commented on src/vm/perfmap.cpp:
```diff
@@ -0,0 +1,141 @@
+    // Build the path to the map file on disk.
+    SString path;
+#ifdef FEATURE_PAL
+    path.Append(W("/tmp/perf-"));
```
Should this use GetTempPath from the PAL instead of hardcoding /tmp/?

janvorli replied:
+1

janvorli commented on src/vm/perfmap.cpp:
```diff
@@ -0,0 +1,141 @@
+        10,     // size in characters
+        10);    // radix
+
+    path.Append((WCHAR *)&strPid);
```
It seems it would be easier to read if you used the SString::Printf method for the whole file name / path formatting.

janvorli commented on src/vm/perfmap.cpp:
```diff
@@ -0,0 +1,141 @@
+    path.Append(W(".map"));
+
+    // Open the file stream.
+    m_FileStream = new (nothrow) CFileStream();
```
Could you please add check for the allocation failure and not call the OpenForWrite on the NULL object pointer.

brianrob commented:
@stephentoub, @janvorli, Thank you for your feedback.

I've fixed the following:

- Use GetTempPathW instead of hardcoding /tmp/.
- Added the NULL check after allocating the file stream.
- #ifdef'd out the contents of perfmap{.h, .cpp}
- Used SString.Printf to clean-up the PerfMap ctor.

I did leave the definition of FEATURE_PERFMAP under the LINUX flag vs. UNIX.  This is because the perfmap feature is only useful on Linux, where the generated file can be used by perf_events.  Perf_events is part of the Linux kernel and does not exist on OSX, FreeBSD, etc, so it isn't usable anywhere but Linux.

mmitche commented on src/vm/prestub.cpp:
```diff
@@ -49,6 +49,10 @@
 #include "stacksampler.h"
 #endif
 
+#ifdef FEATURE_PERFMAP
```
You have some duplicate ifdeffing around here.  Because perfmap.h won't have anything in it unless FEATURE_PERFMAP is specified, this ifdef seems unecessary

brianrob replied:
@mmitche, I've actually removed the ifdefs from perfmap.h and left these in.  I had originally not had ifdefs in perfmap.h, but added them when I was trying to fix the windows build issue I was having.

mmitche commented on src/vm/perfmap.cpp:
```diff
@@ -0,0 +1,139 @@
+//
+
+#ifdef FEATURE_PERFMAP
+#include "common.h"
```
common.h should be above FEATURE_PERFMAP to avoid issues with PCH

brianrob commented:
Thanks everyone for your feedback.  I think I've addressed it all, and have fixed up the build issues as well.

jkotas commented:
LGTM
